### PR TITLE
Add bytesLib and remove memview

### PR DESCRIPTION
**Pull request by AStox**

### CODE REVIEW

File: src/gateway/Gateway.sol [hunk 1]
```diff
@@ -2,7 +2,6 @@
 pragma solidity ^0.8.18;
 pragma abicoder v2;
 
-import {TypedMemView} from "memview-sol/TypedMemView.sol";
```
Comment (offerijns) [inline src/gateway/Gateway.sol:5]: Make sure to also remove the submodule in https://github.com/centrifuge/liquidity-pools/blob/main/.gitmodules

File: src/util/BytesLib.sol [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+ * @title Solidity Bytes Arrays Utils
+ * @author Gonçalo Sá <goncalo.sa@consensys.net>
+ *
+ * @dev Bytes tightly packed arrays utility library for ethereum contracts written in Solidity.
+ *      The library lets you concatenate, slice and type cast bytes arrays both in memory and storage.
+ */
+pragma solidity >=0.8.0 <0.9.0;
+
+library BytesLib {
+    function concat(bytes memory _preBytes, bytes memory _postBytes) internal pure returns (bytes memory) {
```
Comment (offerijns) [inline src/util/BytesLib.sol:12]: I would remove all functions we don't use (like this one, afaik)

**Review by offerijns — COMMENTED** [review]
This looks much cleaner, nic e:)

File: src/util/BytesLib.sol (hunk 2 above)
Comment (AStox) [reply src/util/BytesLib.sol:12]: Final gas savings across `Messages.t.sol`

Test Name | Old Gas Cost | New Gas Cost | Difference
-- | -- | -- | --
testAddCurrency | 12442 | 11396 | 1046
testAddCurrencyEquivalence | 2371 | 1325 | 1046
testAddPool | 4292 | 3491 | 801
testAddPoolEquivalence | 1619 | 818 | 801
testAddTranche | 71545 | 68954 | 2591
testAddTrancheEquivalence | 45904 | 45247 | 657
testAllowPoolCurrency | 9488 | 8176 | 1312
testAllowPoolCurrencyEquivalence | 2615 | 1303 | 1312
testCancelInvestOrder | 23709 | 21896 | 1813
testCancelInvestOrderEquivalence | 3903 | 2090 | 1813
testCancelRedeemOrder | 23719 | 21906 | 1813
testCancelRedeemOrderEquivalence | 3891 | 2078 | 1813
testCollectInvest | 23786 | 21962 | 1824
testCollectInvestEquivalence | 3843 | 2019 | 1824
testCollectRedeem | 23743 | 21919 | 1824
testCollectRedeemEquivalence | 3799 | 1975 | 1824
testDecreaseInvestOrder | 29032 | 26697 | 2335
testDecreaseInvestOrderEquivalence | 4885 | 2550 | 2335
testDecreaseRedeemOrder | 29115 | 26780 | 2335
testDecreaseRedeemOrderEquivalence | 4970 | 2635 | 2335
testExecutedCollectInvest | 38116 | 32446 | 5670
testExecutedCollectInvestEquivalence | 9707 | 4037 | 5670
testExecutedCollectRedeem | 38126 | 32456 | 5670
testExecutedCollectRedeemEquivalence | 9609 | 3939 | 5670
testExecutedDecreaseInvestOrder | 28935 | 26611 | 2324
testExecutedDecreaseInvestOrderEquivalence | 4773 | 2449 | 2324
testExecutedDecreaseRedeemOrder | 28888 | 26564 | 2324
testExecutedDecreaseRedeemOrderEquivalence | 4783 | 2459 | 2324
testFormatDomainAvalanche | 4643 | 4544 | 99
testFormatDomainCentrifuge | 3697 | 3697 | 0
testFormatDomainMoonbaseAlpha | 4731 | 4632 | 99
testFormatDomainMoonbeam | 4666 | 4567 | 99
testIncreaseInvestOrder | 28881 | 26546 | 2335
testIncreaseInvestOrderEquivalence | 4802 | 2467 | 2335
testIncreaseRedeemOrder | 29008 | 26673 | 2335
testIncreaseRedeemOrderEquivalence | 4819 | 2484 | 2335
testTransfer | 32658 | 29277 | 3381
testTransferEquivalence | 6114 | 2733 | 3381
testTransferTrancheTokensToEvm | 39519 | 37607 | 1912
testTransferTrancheTokensToEvmEquivalence | 8511 | 6599 | 1912
testUpdateMember | 21550 | 19737 | 1813
testUpdateMemberEquivalence | 3969 | 2156 | 1813
testUpdateTrancheTokenMetadata | 65323 | 63754 | 1569
testUpdateTrancheTokenMetadataEquivalence | 43548 | 39750 | 3798
testUpdateTrancheTokenPrice | 14450 | 12882 | 1568
testUpdateTrancheTokenPriceEquivalence | 3219 | 1651 | 1568

**Comment by AStox** [comment]
Final gas savings across `Messages.t.sol`



Test Name | Old Gas Cost | New Gas Cost | Difference
-- | -- | -- | --
testAddCurrency | 12442 | 11396 | 1046
testAddCurrencyEquivalence | 2371 | 1325 | 1046
testAddPool | 4292 | 3491 | 801
testAddPoolEquivalence | 1619 | 818 | 801
testAddTranche | 71545 | 68954 | 2591
testAddTrancheEquivalence | 45904 | 45247 | 657
testAllowPoolCurrency | 9488 | 8176 | 1312
testAllowPoolCurrencyEquivalence | 2615 | 1303 | 1312
testCancelInvestOrder | 23709 | 21896 | 1813
testCancelInvestOrderEquivalence | 3903 | 2090 | 1813
testCancelRedeemOrder | 23719 | 21906 | 1813
testCancelRedeemOrderEquivalence | 3891 | 2078 | 1813
testCollectInvest | 23786 | 21962 | 1824
testCollectInvestEquivalence | 3843 | 2019 | 1824
testCollectRedeem | 23743 | 21919 | 1824
testCollectRedeemEquivalence | 3799 | 1975 | 1824
testDecreaseInvestOrder | 29032 | 26697 | 2335
testDecreaseInvestOrderEquivalence | 4885 | 2550 | 2335
testDecreaseRedeemOrder | 29115 | 26780 | 2335
testDecreaseRedeemOrderEquivalence | 4970 | 2635 | 2335
testExecutedCollectInvest | 38116 | 32446 | 5670
testExecutedCollectInvestEquivalence | 9707 | 4037 | 5670
testExecutedCollectRedeem | 38126 | 32456 | 5670
testExecutedCollectRedeemEquivalence | 9609 | 3939 | 5670
testExecutedDecreaseInvestOrder | 28935 | 26611 | 2324
testExecutedDecreaseInvestOrderEquivalence | 4773 | 2449 | 2324
testExecutedDecreaseRedeemOrder | 28888 | 26564 | 2324
testExecutedDecreaseRedeemOrderEquivalence | 4783 | 2459 | 2324
testFormatDomainAvalanche | 4643 | 4544 | 99
testFormatDomainCentrifuge | 3697 | 3697 | 0
testFormatDomainMoonbaseAlpha | 4731 | 4632 | 99
testFormatDomainMoonbeam | 4666 | 4567 | 99
testIncreaseInvestOrder | 28881 | 26546 | 2335
testIncreaseInvestOrderEquivalence | 4802 | 2467 | 2335
testIncreaseRedeemOrder | 29008 | 26673 | 2335
testIncreaseRedeemOrderEquivalence | 4819 | 2484 | 2335
testTransfer | 32658 | 29277 | 3381
testTransferEquivalence | 6114 | 2733 | 3381
testTransferTrancheTokensToEvm | 39519 | 37607 | 1912
testTransferTrancheTokensToEvmEquivalence | 8511 | 6599 | 1912
testUpdateMember | 21550 | 19737 | 1813
testUpdateMemberEquivalence | 3969 | 2156 | 1813
testUpdateTrancheTokenMetadata | 65323 | 63754 | 1569
testUpdateTrancheTokenMetadataEquivalence | 43548 | 39750 | 3798
testUpdateTrancheTokenPrice | 14450 | 12882 | 1568
testUpdateTrancheTokenPriceEquivalence | 3219 | 1651 | 1568

File: src/util/BytesLib.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Unlicense
+/*
+ * @title Solidity Bytes Arrays Utils
+ * @author Gonçalo Sá <goncalo.sa@consensys.net>
+ *
+ * @dev Bytes tightly packed arrays utility library for ethereum contracts written in Solidity.
+ *      The library lets you concatenate, slice and type cast bytes arrays both in memory and storage.
+ */
+pragma solidity >=0.8.0 <0.9.0;
+
```
Comment (offerijns) [inline src/util/BytesLib.sol:10]:
```suggestion
// SPDX-License-Identifier: Unlicense
pragma solidity ^0.8.18;


/*
 * @title Solidity Bytes Arrays Utils
 * @author Gonçalo Sá <goncalo.sa@consensys.net>
 *
 * @dev Bytes tightly packed arrays utility library for ethereum contracts written in Solidity.
 *      The library lets you concatenate, slice and type cast bytes arrays both in memory and storage.
 */
```

**Review by offerijns — APPROVED** [review]

**Pull request merged**